libtenzir/src/series_builder.cpp

```diff
@@ -0,0 +1,1385 @@
+  }
+
+  auto kind() const -> type_kind override {
+    return tag_v<list_type>;
+  }
+
+  auto type() const -> tenzir::type override {
+    return tenzir::type{list_type{elements_.type()}};
+  }
+
+  auto length() const -> int64_t override {
+    return offsets_.length();
+  }
+
+  auto only_null() const -> bool override {
+    return offsets_.null_count() == offsets_.length();
+  }
+
+  auto elements_builder() -> dynamic_builder& {
+    return elements_;
+  }
+
+private:
+  /// This only stores beginning indices. Missing ending offsets are added right
+  /// before the builder is finished.
+  arrow::Int32Builder offsets_;
+
+  dynamic_builder elements_;
+};
+
+template <>
+class typed_builder<record_type> final : public builder_base {
+  friend class detail::field_ref;
+
+public:
+  explicit typed_builder(series_builder_impl* root) : root_{root} {
+  }
+
+  auto append() -> record_ref {
+    length_ += 1;
+    return record_ref{this};
+  }
+
+  auto finish(int64_t count) -> typed_array override {
+    TENZIR_ASSERT_CHEAP(count <= length_);
+    TENZIR_TRACE("finishing {} of {} records with {} fields", count, length(),
+                 fields_.size());
+    auto ty = type();
+    auto field_arrays = std::vector<std::shared_ptr<arrow::Array>>{};
+    field_arrays.reserve(fields_.size());
+    for (auto it = fields_.begin(); it != fields_.end();) {
+      auto& [name, builder] = *it;
+      TENZIR_ASSERT_CHEAP(builder->length() <= length_);
+      if (builder->length() < count) {
+        builder->resize(count);
+      }
+      auto field = builder->finish(count);
+      TENZIR_ASSERT_CHEAP(field.length() == count);
+      field_arrays.push_back(std::move(field.array));
+      auto remove = false;
+      if (builder->length() > 0) {
+        TENZIR_TRACE("not removing field `{}` due to length {}", name,
+                     builder->length());
+      } else {
+        if (builder->is_protected()) {
+          TENZIR_TRACE("not removing field `{}` due to protection", name);
+        } else if (builder.get() == keep_alive_) {
+          TENZIR_TRACE("not removing field `{}` due to keep alive", name);
+        } else {
+          remove = true;
+        }
+      }
+      if (remove) {
+        TENZIR_TRACE("removing field `{}`", name);
+        it = fields_.erase(it);
+      } else {
+        ++it;
+      }
+    }
+    auto null_bitmap = std::shared_ptr<arrow::Buffer>{};
+    if (valid_.length() > 0) {
+      if (count < valid_.length()) {
+        auto total_bits = valid_.length();
+        null_bitmap = valid_.Finish().ValueOrDie();
+        auto copy_bits = total_bits - count;
+        check(valid_.Reserve(copy_bits));
+        valid_.UnsafeAppend(null_bitmap->data(), count, copy_bits);
+        null_bitmap = std::make_shared<arrow::Buffer>(null_bitmap, 0, count);
+      } else {
+        auto missing = count - valid_.length();
+        check(valid_.Append(missing, true));
+        null_bitmap = valid_.Finish().ValueOrDie();
+      }
+    }
```

Comment:
One of the last `TENZIR_TODO()`s was to see whether it can happen that `count < valid_.length()`. Turns out, yes, that is possible. Here is a test: https://github.com/tenzir/tenzir/pull/3503/commits/f99ee8afb6252a740a5d55dcdb292bbf7f09b32e#diff-c08d7dcd3879e83222a3a5e91e68cb1461bd5260a1873a350bfdd7048dd4d5fd

@dominiklohmann  Please take a look and resolve if this makes sense.